### PR TITLE
Authentification

### PR DESCRIPTION
**Pull request by hturnbull93**

### CODE REVIEW

File: bnb.rb [hunk 1]
```diff
@@ -1,13 +1,19 @@
...
   enable :sessions
+  register Sinatra::Flash
 
   get '/' do
-    @user = User.find_by_id(session['user_id'])
+    if session['user_id'] == nil 
+      @user  = nil 
+    else 
+      @user = User.find_by_id(session['user_id'])
+    end
```
Comment (Edward-Phillips) [inline bnb.rb:16]: nice, I like that it checks to see if you're logged in already

File: bnb.rb [hunk 2]
```diff
@@ -29,15 +35,35 @@ class BnB < Sinatra::Base
...
 
-  post '/users' do    
+  post '/users' do
     user = User.create(
-      first_name: params['first_name'], 
-      last_name: params['last_name'], 
-      password: params['password'], 
+      first_name: params['first_name'],
+      last_name: params['last_name'],
+      password: params['password'],
```
Comment (Edward-Phillips) [inline bnb.rb:42]: rubocop approves

File: bnb.rb [hunk 3]
```diff
@@ -29,15 +35,35 @@ class BnB < Sinatra::Base
...
     redirect '/'
   end
 
+  get '/sessions/new' do
+    erb :'sessions/new'
+  end
+
+  post '/sessions' do
+    user = User.authenticate(email: params['email'], password: params['password'])
+    if user != nil
```
Comment (Edward-Phillips) [inline bnb.rb:55]: could this be simplified to if user?

File: lib/user.rb [hunk 4]
```diff
@@ -36,7 +35,21 @@ def self.find_by_id(id)
...
-      password: result[0]['password_digest']
+    )
+  end
+
+  def self.authenticate(email:, password:)
+    connection = PG.connect(dbname: 'bnb_test')
+    result = connection.exec("SELECT * FROM users WHERE email = '#{email}';")
+
+    return unless result.any?
+    return unless BCrypt::Password.new(result[0]['password_digest']) == password
```
Comment (Edward-Phillips) [inline lib/user.rb:46]: nice guards

File: spec/features/authentication_spec.rb [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+  scenario 'user signs out and gets sign  out message' do 
+    existing_user = User.create(first_name: 'john', last_name: 'smith', email: 'johnsmith@smith.com', password: 'johniscool')
+    visit('sessions/new')
+    fill_in('password', with: 'johniscool')
+    fill_in('email', with: 'johnsmith@smith.com')
+    click_on 'login'
+    expect(current_path).to eq '/'
+    click_on 'logout'
+    expect(page).to have_content("You have been signed out")
+    expect(page).to_not have_content("john")
```
Comment (Edward-Phillips) [inline spec/features/authentication_spec.rb:41]: nice and juicy feature test

File: spec/user_spec.rb [hunk 6]
```diff
@@ -25,4 +24,18 @@
...
+    it 'returns user instance if matched email and correct password' do
+      user = User.create(first_name: 'Graham', last_name: 'Falconer', email: 'gman@gmail.com', password: 'password123')
+      auth_user = User.authenticate(email: 'gman@gmail.com', password: 'password123')
+      expect(auth_user.first_name).to eq user.first_name
+    end
+  it 'returns nil if the email is not in the db' do
+      expect(User.authenticate(email: 'gman@gmail.com', password: 'password123')).to eq nil
+    end
+    it 'returns nil if the email exists but password is incorrect' do
+      expect(User.authenticate(email: 'gman@gmail.com', password: 'wrongpassword')).to eq nil
```
Comment (Edward-Phillips) [inline spec/user_spec.rb:38]: user not created in test

File: bnb.rb (hunk 3 above)
Comment (hturnbull93) [reply bnb.rb:55]: done

File: spec/user_spec.rb (hunk 6 above)
Comment (hturnbull93) [reply spec/user_spec.rb]: done

**Pull request merged**